docs/internal_abi.md

```diff
@@ -0,0 +1,321 @@
+If the target does not support hardware floating point then floating point values are stored according to the C ABI for the platform.
+
+### Pointers
+
+We anticipate three common representations for pointers on Verona target systems:
+
+ - 32-bit integers representing an address
+ - 64-bit integers representing an address
+ - CHERI capabilities granting access to an address range.
+
+All Verona objects are at least 8-byte aligned and so the low three bits of a pointer will always be zero.
+When targeting a system that uses integer addresses, all pointers have the same representation.
+When targeting a CHERI system, the following additional rules apply:
+
+ - Pointers to `cown`s are sealed with a type reserved for the runtime.
+   This ensures that `cown`s in immutable objects cannot be dereferenced by unsafe (sandboxed) code and provides some defense in depth against compiler bugs that would accidentally dereference a `cown` pointer.
+ - Pointers to immutable objects are *not* represented by read-only capabilities.
+   Doing so would require expensive operations for reference-count manipulation.
+   When a pointer to an immutable object is passed via the foreign-code layer, it must have the load-storable-capabilities permission removed, such that no reachable capability will ever provide write permission.
+
+*Note:* We currently assume that pointers point to the start of objects.
+Most instruction sets; however, use signed offsets for loads and stores.
+For example, an x86 `mov` instruction can access offsets -128 to 127 in a three-byte instruction and requires (at least) a five-byte instruction for larger displacements.
+It may therefore make sense for object and array pointers to be biased by the displacement of this offset.
+The SPARC ABI, for example, uses this trick to bias the stack pointer to maximise the available displacement in a single instruction.
+
+Pointers to interface types
+---------------------------
+
+Pointers to concrete types are simply pointers.
+Pointers to interface types are a pair of a pointer to the object followed by a pointer to the vtable for the interface.
+There are no additional requirements on alignment for interfaces: the pair of pointers is aligned on the same boundary as a single pointer.
+
+Union types
+-----------
+
+Variables or fields of union types are *discriminated unions*.
+The simplest representation would be to store a buffer of the largest type and an integer large enough to store a unique value for each type.
+For example, a `U8 | Float64 | Foo*` could be represented as a 64-bit buffer followed by a 2-bit integer.
+At an abstract level, the Verona internal ABI does exactly that, with two observations that allow denser packing:
+
+ - Singleton types (equivalent to Pony primitives [*note:* do we have a name for these yet]) have a single instance and so the body can be elided in some cases and the value stored entirely in the tag discriminator.
```

Comment:
"Pony primitives" hasn't been defined here, perhaps a description of "Singleton Types" on a section above this one?

If my interpretation of [this document](https://www.ponylang.io/media/cheatsheet/pony-cheat-sheet.pdf) is correct, a "primitive type" is like a class with only members, which isn't necessarily what I would call "singleton".

Reply:
Yes, it's not quite the same.  The term 'primitive' is confusing in Pony.  They're used as tag values (also overloaded and tag means something else in Pony) where the type is used as a value, rather than as a meaningful holder of state.

Reply:
I honestly can't think of a good name. "Concrete interfaces", "Behaviour class" would all overload existing terms and be even more confusing. I think we can just say "classes without data" or "POBOFP" (plain old bag of function pointers:).

Reply:
@rengolin if I understand what you mean by "members" then no "primitive type"s in Pony aren't like a class with only members.

primitive is a singleton. There are no fields. Only methods.  I think you misread the cheatsheet section.